# RefMasker

**Hard mask homologies between fasta reference sequences identified by Blastn**

---

**Creation : 2015/06/08**

**Last update : 2021/11/16**

---

## Motivation

RefMasker is a **python3.8** object oriented script that was developed in order to attribute more correctly short sequencing reads obtained from a mix of reference sequences whose abundance is highly unbalanced. Indeed, a rare reference with sequence homologies with a much frequent reference can result in possible misattributions of reads to the rarest sequence and thus, a large overestimation of this sequence.

## Principle

1. Users can generate a template configuration file and fill it according to their requirements. The order of references indicated in the configuration file is **CRITICAL** since it will determine the order in which sequences will be masker thereafter.
2. The configuration file containing all program parameters (including reference fasta location) is parsed and verified for validity.
3. Following the order indicated in the configuration file, reference fasta files are uncompressed (if needed), parsed, and indexed using a memory mapping.
4. An iterative masking is performed, starting from the last reference (**subject**) against all the references listed before (**queries**). For each new iteration the penultimate reference from the previous iteration becomes the **subject** and is removed from the **queries** (see figure below).
5. When the list of queries is empty the iteration stops.
6. Depending of the user requirements, blast and masking reports are generated.

![RefMasker_iteration](https://raw.githubusercontent.com/a-slide/RefMasker/master/fig/RefMasker_iteration.png)

**Details of iterations**
    
* Imperfect matches between the subject and the queries are found using a wrapper of NCBI Blast+ ([pyBlast submodule](http://a-slide.github.io/pyBlast))
* If matches were found, the program writes a masked version of the subject reference where each positions of the subject overlapping hits is replaced by a 'N' base (hard masking).
* The subject reference is removed from the reference list.

## Get and install

Clone the repository using `git clone https://github.com/emlec/RefMasker.git`.
You may proceed using Singularity or Conda.

### Singularity

* Please install Singularity, if not already installed.
You can then build RefMasker container using `sudo singularity build refmasker singularity/Singularity.rcp`

* Refmasker can be used from within the container.
To do so, start a shell in the container using `singularity shell refmasker`.

### Conda

* Please install Conda, if not already installed.
You can then create RefMasker conda environment using `conda env create -f conda/conda.yaml`

* Activate the environment using `conda activate RefMasker`. You can then use RefMasker.

## Usage

In the folder where files will be created:

```
Usage: RefMasker.py -c Conf.txt [-i -h]

Options:
  --version     show program's version number and exit
  -h, --help    show this help message and exit
  -c CONF_FILE  Path to the configuration file [Mandatory]
  -i            Generate an example configuration file and exit [Facultative]
```
  
An example configuration file can be generated by running the program with the option -i

The possible options are extensively described in the configuration file.

The program can be tested from the test folder with the dataset provided and the default configuration file.

```
cd ./test/result
RefMasker.py -i
RefMasker.py -c Quade_conf_file.txt
```

## Authors and Contact

* Adrien Leger <aleg@ebi.ac.uk> @a-slide
* Emilie Lecomte <emilie.lecomte@univ-nantes.fr> @emlec
