# RefMasker

**Hard mask homologies between fasta reference sequences identified by blast**

---

**Creation : 2015/06/08**

**Last update : 2015/06/08**

---

## Motivation

RefMasker is a **python2.7** object oriented script that was developed in order to attribute more correctly short sequencing reads obtained from a mix of sequences whose abundance is highly unbalance. Indeed, a rare sequence with large sequence homologies with a much frequent one can results in possible misattribution of reads to the rarest sequence and thus, a large overestimation of this sequence.
This program can find sequence homologies of a list of reference fasta files. The order of references is **CRITICAL** since the last reference will be masked by all the others, then the penultimate will masked by all the sequence listed before and and so on until there is only one reference remaining.

![RefMasker_iteration](https://raw.githubusercontent.com/a-slide/RefMasker/master/fig/RefMasker_iteration.svg)

## Principle

1. A configuration file containing all program parameters (including reference fasta location) is parsed and verified for validity.
2. Reference fasta references are parsed ..............................
3. 

## Dependencies

The program was developed under Linux Mint 17 and was not tested with other OS.
In addition to python2.7 the following dependencies are required for proper program execution:

* python package [pyfasta](https://github.com/brentp/pyfasta/) 0.5.2 +
* python package [pytest](

If you have pip already installed, enter the following line to install packages: ```sudo pip install pyfasta```


## Get and install Quade

* Clone the repository in **recursive mode** to download the main repo and its submodules ```git clone --recursive https://github.com/a-slide/RefMasker.git```

* Enter the src folder of the program folder and make the main script executable ```sudo chmod u+x RefMasker.py```

* Finally, add RefMasker.py to your PATH

## Usage

In the folder where files will be created
    
    Usage: RefMasker.py -c Conf.txt [-i -h]
    
    Options:
      --version     show program's version number and exit
      -h, --help    show this help message and exit
      -c CONF_FILE  Path to the configuration file [Mandatory]
      -i            Generate an example configuration file and exit [Facultative]
      
An example configuration file can be generated by running the program with the option -i
The possible options are extensively described in the configuration file.
The program can be tested from the test folder with the dataset provided and the default configuration file.
```
cd ./test/result
RefMasker.py -i
RefMasker.py -c Quade_conf_file.txt

```
## Testings

Unit testing can be performed

## Authors and Contact

Adrien Leger - 2015

* <adrien.leger@gmail.com> - <adrien.leger@inserm.fr> - <adrien.leger@univ-nantes.fr>
* [Github](https://github.com/a-slide)
* [Atlantic Gene Therapies - INSERM 1089](http://www.atlantic-gene-therapies.fr/)
* [Drowned in genomics](http://a-slide.github.io/)
